go/fn/doc.go

```diff
@@ -13,58 +13,66 @@
-}
-
-// Add your function logic in run.
-// `resourcelist.functionConfig` is assigned to YourKRMFn object. Your functionConfig kind shall be `YourKRMFn` or `ConfigMap` (limited usage)
-// `resourcelist.items` is the passed-in `items` parameter.
-// `resourcellist.result` is empty, you can add result via `ctx` methods like `AddGeneralResult`, `AddErrResultAndDie`.
-func (r *YourKRMFn) Run(ctx *fn.Context, functionConfig *fn.KubeObject, items []*fn.KubeObject) {
-	for _, o := range items {
-		// Your code
-	}
-}
-
-func main() {
-	if err := fn.AsMain(&YourKRMFn{}); err != nil {
-		os.Exit(1)
-	}
-}
-```
+Read more about how to use KRM functions in https://kpt.dev/book/04-using-functions/
+
+Read more about how to develop a KRM function in https://kpt.dev/book/05-developing-functions/
+
+A general workflow is:
+  1. Reads the "ResourceList" object from STDIN.
+  2. Gets the function configs from the "ResourceList.FunctionConfig".
+  3. Mutate or validate the Kubernetes YAML resources from the "ResourceList.Items" field with the function configs.
+  4. Writes the modified "ResourceList" to STDOUT.
+
+KubeObject
+
+The KubeObject is the basic unit to perform operations on KRM resources.
+
+In the AsMain, both "Items" and "FunctionConfigs"
+are converted to the KubeObject(s).
+
+If you are familiar with unstructured.Unstructured, using KubeObject is as simple as using unstructured.Unstructured.
+You can call function like `NestedStringOrDie` `SetNestedStringMap`, etc.
+
+Except that KubeObject will not have pass-in interface arguments, nor will return an interface.
+Instead, you shall treat each KubeObject field (slice, or non-string map）as SubObject.
+
+SubObject also have most of the KubeObject methods, except the MetaType or NameType specific methods like "GetNamespace", "SetLabel".
+This is because SubObject is designed as a sub object of KubeObject. SubObject to KubeObject is like `spec` section to `Deployment`.
+You can get the Deployment name from `metadata.name`, KubeObject.GetName() or KubeObject.NestedString("metadata", "name").
+But you cannot get "metadata.name" from a Deployment "spec". For "spec" SubObject, you can get the ".replicas" field by
+SubObject.NestedInt64("replicas")
+
+Besides unstructured style, another way to use KubeObject is to purely work on the KubeObject/SubObject by calling
+"GetMap", "GetSlice", "UpsertMap" which expects the return to be SubObject(s) pointer.
+
+AsMain
+
+"AsMain" is the main entrypoint. In most cases, you only need to provide the mutator or validation logic and have AsMain
+handles the ResourceList parsing, KRM resource field type detection, read from STDIN and write to STDOUT.
+
+"AsMain" accepts a struct that either implement the ResourceListProcessor interface or Runner interface.
+
+See github.com/GoogleContainerTools/kpt-functions-sdk/go/fn/examples for detailed usage.
+
 */
-package fn
+package fn // import "github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
```

Comment:
Why add this?

Reply:
oh, this is for import path checking. But I think it's disabled by go modules already.

Reply:
cleaned up